// mutable/dist/test/enum.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
var mu = require("../src");
var test_drivers_1 = require("../test-kit/test-drivers");
describe('Enum Type', function () {
    it('exists on the mu object', function () {
        chai_1.expect(mu.EnumBase).to.be.a("function");
    });
    describe("instantiation", function () {
        it("should return a class", function () {
            var Ape = mu.defineEnum(["chimp", "gorilla"]);
            chai_1.expect(Ape).to.be.a("function");
        });
        it("enum cannot be instantiated", function () {
            var Ape = mu.defineEnum(["chimp", "gorilla"]);
            chai_1.expect(function () {
                new Ape();
            }).to.report({ level: /error/ });
        });
        it("enum extends EnumBase", function () {
            var Ape = mu.defineEnum(["chimp", "gorilla"]);
            chai_1.expect(Ape.chimp).to.be.instanceof(mu.EnumBase);
        });
        it("enum field can be generic by being EnumBase", function () {
            var Ape = mu.defineEnum(["chimp", "gorilla"]);
            var Dog = mu.defineEnum(["poodle", "lab"]);
            var TestType = test_drivers_1.aDataTypeWithSpec({
                ape: mu.EnumBase
            });
            var tt = TestType.create({ ape: Ape.chimp });
            chai_1.expect(tt.ape).to.be.equal(Ape.chimp);
            tt = TestType.create({ ape: Dog.poodle });
            chai_1.expect(tt.ape).to.be.equal(Dog.poodle);
        });
        it("generic enum has static validate method", function () {
            var Ape = mu.defineEnum(["chimp", "gorilla"]);
            chai_1.expect(mu.EnumBase.validate(Ape.chimp)).to.be.ok;
            chai_1.expect(mu.EnumBase.validate("sugar booger")).to.not.be.ok;
        });
        it("enum validate method only allows own instance", function () {
            var Ape = mu.defineEnum(["chimp", "gorilla"]);
            chai_1.expect(Ape.validate(Ape.chimp)).to.be.ok;
            var Dog = mu.defineEnum(["poodle", "lab"]);
            chai_1.expect(Ape.validate(Dog.poodle)).to.not.be.ok;
        });
        it("enum can be initialized", function () {
            var Ape = mu.defineEnum(["chimp", "gorilla"]);
            var ape = Ape.chimp;
            chai_1.expect(ape).to.be.equal(Ape.chimp);
        });
        it("should return value string for toString()", function () {
            var Ape = mu.defineEnum(["chimp", "gorilla"]);
            chai_1.expect(Ape.chimp.toString()).to.be.equal("chimp");
            chai_1.expect(Ape.gorilla.toString()).to.be.equal("gorilla");
        });
        it("enum can be initialized as member in a custom type", function () {
            var Ape = mu.defineEnum(["chimp", "gorilla"]);
            var TestType = test_drivers_1.aDataTypeWithSpec({
                ape: Ape
            });
            var tt = TestType.create(Ape.chimp);
            chai_1.expect(tt.ape).to.be.equal(Ape.chimp);
        });
        it("enum can be initialized using withDefault", function () {
            var Ape = mu.defineEnum(["chimp", "gorilla"]);
            var TestType = test_drivers_1.aDataTypeWithSpec({
                ape: Ape.withDefault(Ape.gorilla)
            });
            var tt = TestType.create();
            chai_1.expect(tt.ape).to.be.equal(Ape.gorilla);
        });
        it("enum can receive value from create of parent type", function () {
            var Ape = mu.defineEnum(["chimp", "gorilla"]);
            var TestType = test_drivers_1.aDataTypeWithSpec({
                ape: Ape.withDefault(Ape.gorilla)
            });
            var tt = TestType.create({ ape: Ape.chimp });
            chai_1.expect(tt.ape).to.be.equal(Ape.chimp);
        });
        it("each member can have a string value", function () {
            var Ape = mu.defineEnum(["chimp", "gorilla"]);
            chai_1.expect(Ape.chimp.value).to.be.equal("chimp");
            chai_1.expect(Ape.gorilla.value).to.be.equal("gorilla");
        });
        it("members can have a custom string value", function () {
            var Ape = mu.defineEnum({ chimp: "_chimp_", gorilla: "_gorilla_" });
            chai_1.expect("" + Ape.chimp).to.be.equal("_chimp_");
            chai_1.expect("" + Ape.gorilla).to.be.equal("_gorilla_");
        });
        it("members can have number value", function () {
            var Ape = mu.defineEnum({ chimp: 0, gorilla: 1 });
            chai_1.expect(Ape.chimp.value).to.be.equal(0);
            chai_1.expect(Ape.gorilla.value).to.be.equal(1);
        });
        it("members can have object values", function () {
            var Ape = mu.defineEnum({
                chimp: { dateOfBirth: "11/11/1980" },
                gorilla: { dateOfBirth: "12/12/1995" }
            });
            chai_1.expect(Ape.chimp.value.dateOfBirth).to.be.equal("11/11/1980");
            chai_1.expect(Ape.gorilla.value.dateOfBirth).to.be.equal("12/12/1995");
        });
        it("member is immutable", function () {
            var Ape = mu.defineEnum({ chimp: 1, gorilla: 2 });
            chai_1.expect(function () {
                Ape.chimp.blyat = "blyat";
            }).to.throw('object is not extensible');
        });
        it("instanceOf works", function () {
            var Ape = mu.defineEnum({ chimp: 1, gorilla: 2 });
            chai_1.expect(Ape.chimp).to.be.instanceof(Ape);
        });
        it("members have a key which is the originl key in the def", function () {
            var Ape = mu.defineEnum({ chimp: 1, gorilla: 2 });
            chai_1.expect(Ape.chimp.key).to.be.equal("chimp");
        });
        it("should return current value", function () {
            var ImageSizing = mu.defineEnum({
                NONE: null,
                COVER: "cover",
                CONTAIN: "contain"
            });
            chai_1.expect(ImageSizing.NONE.toJSON()).to.be.null;
            chai_1.expect(ImageSizing.COVER.toJSON()).to.be.equal("cover");
            chai_1.expect(ImageSizing.CONTAIN.toJSON()).to.be.equal("contain");
        });
        it("should initiate from value string", function () {
            var ImageSizing = mu.defineEnum({
                NONE: null,
                COVER: "cover",
                CONTAIN: "contain"
            });
            var enumIns = ImageSizing.create(ImageSizing.COVER.toJSON());
            chai_1.expect(enumIns).to.be.equal(ImageSizing.COVER);
        });
        it("should initiate from value string as part of complex data", function () {
            var ImageSizing = mu.defineEnum({
                NONE: null,
                COVER: "cover",
                CONTAIN: "contain"
            });
            var Complex = test_drivers_1.aDataTypeWithSpec({
                size: ImageSizing
            }, 'User');
            var complex = new Complex({ size: "cover" });
            chai_1.expect(complex.size).to.be.equal(ImageSizing.COVER);
        });
        it("should initiate from default value as part of complex data when no initial value is provided", function () {
            var ImageSizing = mu.defineEnum({
                NONE: null,
                COVER: "cover",
                CONTAIN: "contain"
            });
            var Complex = test_drivers_1.aDataTypeWithSpec({
                size: ImageSizing.withDefault(ImageSizing.CONTAIN)
            }, 'User');
            var complex = new Complex({ size: "contain" });
            chai_1.expect(complex.size).to.be.equal(ImageSizing.CONTAIN);
        });
        it("should not be dirtyable", function () {
            var ImageSizing = mu.defineEnum(["A", "B"]);
            chai_1.expect(ImageSizing.$isDirtyable).to.not.be.defined;
            chai_1.expect(ImageSizing.A.$isDirtyable).to.not.be.defined;
            chai_1.expect(ImageSizing.B.$isDirtyable).to.not.be.defined;
        });
    });
    describe('toJSON', function () {
        it('should return the enum value that can initiate the value', function () {
            var ApeWithValue = mu.defineEnum({
                'CHIMP': 'small ape',
                'GORILLA': 'large ape',
            });
            var res = ApeWithValue.GORILLA.toJSON();
            var deserializeApe = new ApeWithValue(res);
            chai_1.expect(res).to.equal('large ape');
            chai_1.expect(deserializeApe).to.equal(ApeWithValue.GORILLA);
        });
    });
    describe('toJS', function () {
        it('should return the enum value that can initiate the value', function () {
            var ApeWithValue = mu.defineEnum({
                'CHIMP': 'small ape',
                'GORILLA': 'large ape',
            });
            var res = ApeWithValue.GORILLA.toJS();
            var deserializeApe = new ApeWithValue(res);
            chai_1.expect(res).to.equal('large ape');
            chai_1.expect(deserializeApe).to.equal(ApeWithValue.GORILLA);
        });
    });
});
