// mutable/dist/test/es5-map/builders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("lodash");
var mu = require("../../src");
var test_drivers_1 = require("../../test-kit/test-drivers");
function aNumberMap(optionalMap) {
    return mu.Es5Map.of(mu.Number).create(optionalMap || {});
}
exports.aNumberMap = aNumberMap;
function aUserTypeMap(optionalMap) {
    return mu.Es5Map.of(exports.UserType).create(optionalMap || []);
}
exports.aUserTypeMap = aUserTypeMap;
function aUnionTypeMap(optionalMap) {
    return mu.Es5Map.of(mu.either(exports.CheeseType, exports.UserType)).create(optionalMap || []);
}
exports.aUnionTypeMap = aUnionTypeMap;
exports.UserType = test_drivers_1.aDataTypeWithSpec({
    name: mu.String.withDefault(''),
    age: mu.Number.withDefault(10)
}, 'User');
exports.CheeseType = test_drivers_1.aDataTypeWithSpec({
    name: mu.String.withDefault(''),
    age: mu.Number.withDefault(10)
}, 'Cheese');
var exported = { CheeseType: exports.CheeseType, UserType: exports.UserType, aNumberMap: aNumberMap, aUserTypeMap: aUserTypeMap, aUnionTypeMap: aUnionTypeMap };
function asReadOnly() {
    return _.mapValues(exported, function (prop) {
        if (prop.id) {
            return prop;
        }
        else {
            return _.flow(prop, function (list) { return list.$asReadOnly(); });
        }
    });
}
exports.asReadOnly = asReadOnly;
