// mutable/dist/test/list/mutable/instance.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
var mu = require("../../../src");
var builders_1 = require("../builders");
describe('List', function () {
    describe('mu instance', function () {
        it('Should have default length', function () {
            var numberList = builders_1.aNumberList([1, 2, 3]);
            chai_1.expect(numberList.length).to.equal(3);
        });
        describe('as field on data object', function () {
            var GroupType;
            beforeEach(function () {
                GroupType = mu.define('GroupType', {
                    spec: function () {
                        return {
                            title: mu.String,
                            users: mu.List.of(builders_1.UserType)
                        };
                    }
                });
            });
            it('Should be modified from json ', function () {
                var groupData = new GroupType();
                groupData.users = mu.List.of(builders_1.UserType).create([
                    { 'name': 'tom', 'age': 25 },
                    { 'name': 'omri', 'age': 35 }
                ]);
                chai_1.expect(groupData.users.at(0).name).to.equal('tom');
                chai_1.expect(groupData.users.at(0).age).to.equal(25);
                chai_1.expect(groupData.users.at(1).name).to.equal('omri');
                chai_1.expect(groupData.users.at(1).age).to.equal(35);
            });
        });
    });
});
