// mutable/dist/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var config_1 = require("./config");
exports.config = config_1.default;
var define_1 = require("./objects/define");
exports.define = define_1.defineClass;
var object_1 = require("./objects/object");
exports.Object = object_1.MuObject;
var any_1 = require("./data-types/any");
exports.Any = any_1.Any;
var string_1 = require("./data-types/primitives/string");
exports.String = string_1.default;
var boolean_1 = require("./data-types/primitives/boolean");
exports.Boolean = boolean_1.default;
var number_1 = require("./data-types/primitives/number");
exports.Number = number_1.default;
var function_1 = require("./data-types/primitives/function");
exports.Function = function_1.default;
var reference_1 = require("./objects/reference");
exports.Reference = reference_1.default;
var validation_1 = require("./core/validation");
exports.validation = validation_1.default;
var list_1 = require("./data-types/list");
exports.List = list_1.default;
var es5_map_1 = require("./data-types/es5-map");
exports.Es5Map = es5_map_1.default;
exports.Map = es5_map_1.default;
var utils_1 = require("./utils");
exports.getReadableValueTypeName = utils_1.getReadableValueTypeName;
var base_1 = require("./core/base");
exports.Base = base_1.MuBase;
var props_base_1 = require("./props-base");
exports.PropsBase = props_base_1.propsBase;
var lifecycle_1 = require("./core/lifecycle");
exports.LifeCycleManager = lifecycle_1.LifeCycleManager;
var define_enum_1 = require("./data-types/define-enum");
exports.defineEnum = define_enum_1.defineEnum;
exports.EnumBase = define_enum_1.EnumBase;
var generic_types_1 = require("./core/generic-types");
exports.either = generic_types_1.either;
var computed_decorator_1 = require("./core/computed-decorator");
exports.computed = computed_decorator_1.computed;
var ifndef_1 = require("ifndef");
var object_2 = require("./objects/object");
var base_2 = require("./core/base");
var types_1 = require("./types");
// TODO export isType itself
function isAny(type) {
    return types_1.isType(type);
}
exports.isAny = isAny;
function isMutable(type) {
    return type && type.prototype instanceof base_2.MuBase;
}
exports.isMutable = isMutable;
function isClass(type) {
    return type && type.prototype instanceof object_2.MuObject;
}
exports.isClass = isClass;
function isEnum(type) {
    return type && type.name === 'EnumType';
}
exports.isEnum = isEnum;
function isNullable(type) {
    return !!(type && type.options && type.options.nullable);
}
exports.isNullable = isNullable;
module.exports = ifndef_1.ifndef('mutable', module.exports);
